File: fortran/dust.f90
module dust
    use misc
    use photon_packet

    implicit none
    private
    public :: dust_grain, init_dust, clean_dust, planck_mean_opacity
    public :: dust_emit, dust_absorb, dust_isoscatt, opacity, random_nu
    public :: albedo, int_dBnu_knu

    type dust_grain
        sequence
        real(8), allocatable, dimension(:) :: nu, lam, kabs, ksca, kext, albedo
        real(8), allocatable, dimension(:) :: temp, planck_opacity
        real(8), allocatable, dimension(:) :: int_dBnu_knu
        real(8), allocatable, dimension(:) :: dplanck_opacity_dT, &
            &dint_dBnu_knu_DT
        real(8), allocatable, dimension(:,:) :: Bnu, dBnu, dBnudT, ddBnudT
        integer :: nlam
    end type dust_grain

contains
    
    subroutine init_dust(d,i)
        implicit none
        type(dust_grain) :: d
        real(8), allocatable, dimension(:) :: lam, kabs, ksca, dkextdnu, &
            &dalbedodnu
        real(8), allocatable, dimension(:) :: temp, planck_opacity, int_dBnu_knu
        real(8), allocatable, dimension(:) :: dplanck_opacity_dT, &
            &dint_dBnu_knu_DT
        real(8), allocatable, dimension(:,:) :: Bnu, dBnu, dBnudT, ddBnudT
        integer :: i, j, k, dummy, nlam, nspecies, ntemp
        character*512 :: dust_file

        open(unit=2,file="dust.dat")
        read(2,*) nspecies
        do j=1, i
            read(2,*) dust_file
        end do
        close(2)

        open(unit=2,file=trim(dust_file))
        read(2,*) dummy
        read(2,*) nlam

        d%nlam = nlam

        allocate (lam(nlam))
        allocate (kabs(nlam))
        allocate (ksca(nlam))
        allocate (dkextdnu(nlam))
        allocate (dalbedodnu(nlam))
        allocate (d%lam(nlam))
        allocate (d%nu(nlam))
        allocate (d%kabs(nlam))
        allocate (d%ksca(nlam))
        allocate (d%kext(nlam))
        allocate (d%albedo(nlam))

        do j=1, nlam
            read(2,*) lam(j), kabs(j), ksca(j)
        end do

        close(2)

        d%lam = lam*1.0e-4
        d%nu = c_l/(lam*1.0e-4)
        d%kabs = kabs
        d%ksca = ksca
        d%kext = kabs+ksca
        d%albedo = ksca/(kabs+ksca)

        open(unit=2,file="lookup_table3.txt")
        do j=1, nlam
            read(2,*) dkextdnu(j), dalbedodnu(j)
        end do
        close(2)

        deallocate (lam)
        deallocate (kabs)
        deallocate (ksca)
        deallocate (dkextdnu)
        deallocate (dalbedodnu)

        open(unit=2,file="lookup_table1.txt")
        read(2,*) ntemp

        allocate (d%temp(ntemp))
        allocate (d%planck_opacity(ntemp))
        allocate (d%int_dBnu_knu(ntemp))
        allocate (d%dplanck_opacity_dT(ntemp))
        allocate (d%dint_dBnu_knu_dT(ntemp))
        allocate (temp(ntemp))
        allocate (planck_opacity(ntemp))
        allocate (int_dBnu_knu(ntemp))
        allocate (dplanck_opacity_dT(ntemp))
        allocate (dint_dBnu_knu_dT(ntemp))

        do j=1, ntemp
            read(2,*) temp(j), planck_opacity(j), int_dBnu_knu(j), &
                &dplanck_opacity_dT(j), dint_dBnu_knu_dT(j)
        end do
        d%temp = temp
        d%planck_opacity = planck_opacity
        d%int_dBnu_knu = int_dBnu_knu
        d%dplanck_opacity_dT = dplanck_opacity_dT
        d%dint_dBnu_knu_dT = dint_dBnu_knu_dT

        close(2)

        allocate (d%Bnu(ntemp,nlam))
        allocate (d%dBnu(ntemp,nlam))
        allocate (d%dBnudT(ntemp,nlam))
        allocate (d%ddBnudT(ntemp,nlam))
        allocate (Bnu(ntemp,nlam))
        allocate (dBnu(ntemp,nlam))
        allocate (dBnudT(ntemp,nlam))
        allocate (ddBnudT(ntemp,nlam))

        open(unit=2,file="lookup_table2.txt")

        do j=1, ntemp
            do k=1, nlam
                read(2,*), Bnu(j,k), dBnu(j,k), dBnudT(j,k), ddBnudT(j,k)
            enddo
        enddo

        close(2)

        d%Bnu = Bnu
        d%dBnu = dBnu
        d%dBnudT = dBnudT
        d%ddBnudT = ddBnudT

        deallocate (temp, planck_opacity, int_dBnu_knu, dplanck_opacity_dT,&
            &dint_dBnu_knu_dT)
        deallocate(Bnu)
        deallocate(dBnu)
        deallocate(dBnudT)
        deallocate(ddBnudT)

    end subroutine init_dust

    subroutine dust_emit(d,p,x,y,z,T,bw,species)
        implicit none
        type(photon) :: p
        type(dust_grain) :: d
        real(8) :: x, y, z, T, cost, sint, phi, sinp, cosp
        logical :: bw
        type(dust_grain), allocatable, dimension(:) :: species
        integer :: i, nspecies

        p%x = x
        p%y = y
        p%z = z

        cost = 2*random()-1
        sint = sqrt(1.-cost**2)
        phi = 2.*3.1415927*random()
        sinp = sin(phi)
        cosp = cos(phi)

        p%nx = sint*cosp
        p%ny = sint*sinp
        p%nz = cost

        p%invnx = 1.0d0/p%nx
        p%invny = 1.0d0/p%ny
        p%invnz = 1.0d0/p%nz

        call random_nu(d,T,p,bw)

        nspecies = size(species)

        allocate (p%current_kext(nspecies))
        allocate (p%current_albedo(nspecies))

        do i=1, nspecies
            p%current_kext(i) = opacity(species(i),p%nu)
            p%current_albedo(i) = albedo(species(i),p%nu)
        end do
    end subroutine dust_emit

    subroutine dust_absorb(d,p,T,bw,species)
        implicit none
        type(photon) :: p
        type(dust_grain) :: d
        real(8) :: T, cost, sint, phi, cosp, sinp
        logical :: bw
        type(dust_grain), dimension(:) :: species
        integer :: i

        cost = 2*random()-1.
        sint = sqrt(1.-cost**2)
        phi = 2.*3.1415927*random()
        cosp = cos(phi)
        sinp = sin(phi)

        p%nx = sint*cosp
        p%ny = sint*sinp
        p%nz = cost

        p%invnx = 1.0d0/p%nx
        p%invny = 1.0d0/p%ny
        p%invnz = 1.0d0/p%nz

        call random_nu(d,T,p,bw)

        do i=1, size(species)
            p%current_kext(i) = opacity(d,p%nu)
            p%current_albedo(i) = albedo(d,p%nu)
        end do
    end subroutine dust_absorb

    subroutine dust_isoscatt(p)
        implicit none
        type(photon) :: p
        real(8) :: cost, sint, phi, cosp, sinp

        cost = 2*random()-1.
        sint = sqrt(1.-cost**2)
        phi = 2.*3.1415927*random()
        cosp = cos(phi)
        sinp = sin(phi)

        p%nx = sint*cosp
        p%ny = sint*sinp
        p%nz = cost

        p%invnx = 1.0d0/p%nx
        p%invny = 1.0d0/p%ny
        p%invnz = 1.0d0/p%nz
    end subroutine dust_isoscatt

    subroutine random_nu(d,T,p,bw)
        implicit none
        type(photon) :: p
        type(dust_grain) :: d
        real(8) :: nu, Prob, T, norm, ksi, tot
        real(8), allocatable, dimension(:) :: F
        integer :: i, n
        logical :: bw

        n = size(d%nu)

        allocate (F(n))
        if (bw) then
            call dBnu_arr(d,T,n,F)
            norm = 1.0d0/int_dBnu_knu(d,T)
        else
            !call B_nu_arr(T,d%nu,n,F)
            call Bnu_arr(d,T,n,F)
            norm = -pi/(planck_mean_opacity(d,T)*sigma*T**4)
        end if

        ksi = random()

        tot = 0.0d0
        do i=1, n-1
            tot = tot + 0.5d0*(d%nu(n-i+1)-d%nu(n-i))*&
                &(d%kext(n-i)*F(n-i)+d%kext(n-i+1)*F(n-i+1))
            Prob = tot*norm

            if (Prob.gt.ksi) then
                p%nu = random()*(d%nu(n-i+1)-d%nu(n-i))+&
                    &d%nu(n-i)
                return
            end if
        end do
        deallocate (F)

    end subroutine random_nu

    real(8) function opacity(d,nu)
        implicit none
        type(dust_grain) :: d
        real(8) :: nu
        integer :: l

        call find_freq_bin(d,nu,l)

        opacity = (d%kext(l+1)-d%kext(l))/(d%nu(l+1)-d%nu(l))*(nu-d%nu(l))+&
            &d%kext(l)

    end function opacity

    real(8) function albedo(d,nu)
        implicit none
        type(dust_grain) :: d
        real(8) :: nu
        integer :: l

        call find_freq_bin(d,nu,l)

        albedo = (d%albedo(l+1)-d%albedo(l))/(d%nu(l+1)-d%nu(l))*(nu-d%nu(l))+&
            &d%albedo(l)

    end function albedo

    real(8) function planck_mean_opacity(d,T)
        implicit none
        type(dust_grain) :: d
        real(8) :: T
        integer :: n

        n = find_in_arr(T,d%temp)

        planck_mean_opacity = d%dplanck_opacity_dT(n)*(T-d%temp(n))+&
            &d%planck_opacity(n)

    end function planck_mean_opacity

    real(8) function int_dBnu_knu(d,T)
        implicit none
        type(dust_grain) :: d
        real(8) :: T
        integer :: n

        n = find_in_arr(T,d%temp)

        int_dBnu_knu = d%dint_dBnu_knu_dT(n)*(T-d%temp(n))+d%int_dBnu_knu(n)

    end function int_dBnu_knu

    subroutine Bnu_arr(d,T,n,Bnu)
        implicit none
        type(dust_grain) :: d
        real(8) :: T
        integer :: n, i
        real(8), dimension(n) :: Bnu

        i = find_in_arr(T,d%temp)

        Bnu = d%dBnudT(i,:)*(T-d%temp(i))+d%Bnu(i,:)

    end subroutine Bnu_arr

    subroutine dBnu_arr(d,T,n,dBnu)
        implicit none
        type(dust_grain) :: d
        real(8) :: T
        integer :: n, i
        real(8), dimension(n) :: dBnu

        i = find_in_arr(T,d%temp)

        dBnu = d%ddBnudT(i,:)*(T-d%temp(i))+d%dBnu(i,:)

    end subroutine dBnu_arr

    subroutine find_freq_bin(d,nu,l)
        implicit none
        type(dust_grain) :: d
        real(8) :: nu
        integer :: l, i, lmin, lmax, ltest
        logical :: not_found

        lmin = 1
        lmax = size(d%nu)
        not_found = .true.
        i = 1
        do while (not_found.and.(i.le.10))
            ltest = (lmax-lmin)/2+lmin

            if (in_freq_bin(d,nu,ltest)) then
                l = ltest
                not_found = .false.
            else
                if (nu.lt.d%nu(ltest)) then
                    lmin = ltest
                else
                    lmax = ltest
                endif
            endif
            i = i+1
        end do
    end subroutine find_freq_bin

    logical function in_freq_bin(d,nu,l)
        implicit none
        type(dust_grain) :: d
        real(8) :: nu
        integer :: l

        if ((nu.le.d%nu(l)).and.(nu.gt.d%nu(l+1))) then
            in_freq_bin = .true.
        else
            in_freq_bin = .false.
        end if
    end function in_freq_bin

    subroutine clean_dust(d)
        implicit none
        type(dust_grain) :: d

        deallocate(d%lam)
        deallocate(d%nu)
        deallocate(d%kabs)
        deallocate(d%ksca)
        deallocate(d%kext)
        deallocate(d%albedo)
        deallocate(d%temp)
        deallocate(d%planck_opacity)
        deallocate(d%int_dBnu_knu)
        deallocate(d%dplanck_opacity_dT)
        deallocate(d%dint_dBnu_knu_dT)
        deallocate(d%Bnu)
        deallocate(d%dBnu)
        deallocate(d%dBnudT)
        deallocate(d%ddBnudT)

    end subroutine clean_dust

end module dust

File: src/dust.cc
#include "dust.h"

/* Functions to set up the dust. */

Dust::Dust(py::array_t<double> __lam, py::array_t<double> __kabs,
            py::array_t<double> __ksca) {

    _lam = __lam; _kabs = __kabs; _ksca = __ksca;

    // Load the array buffers to get the proper setup info.

    auto _lam_buf = __lam.request(); auto _kabs_buf = __kabs.request();
    auto _ksca_buf = __ksca.request();

    if (_lam_buf.ndim != 1 || _kabs_buf.ndim != 1 || _ksca_buf.ndim != 1)
        throw std::runtime_error("Number of dimensions must be one");

    // Now get the correct format.

    nlam = _lam_buf.shape[0];
    lam = (double *) _lam_buf.ptr;
    kabs = (double *) _kabs_buf.ptr;
    ksca = (double *) _ksca_buf.ptr;

    // Set up the volume of each cell.

    _nu = py::array_t<double>(nlam);
    _kext = py::array_t<double>(nlam);
    _albedo = py::array_t<double>(nlam);

    auto _nu_buf = _nu.request(); nu = (double *) _nu_buf.ptr;
    auto _kext_buf = _kext.request(); kext = (double *) _kext_buf.ptr;
    auto _albedo_buf = _albedo.request(); albedo = (double *) _albedo_buf.ptr;

    // Finally, calculate their values.

    for (int i = 0; i < nlam; i++) {
        nu[i] = c_l / lam[i]; 
        kext[i] = kabs[i] + ksca[i];
        albedo[i] = ksca[i] / kext[i];
    }

    // Catch if nu is not increasing.

    if (nu[1] - nu[0] <= 0.)
        throw std::runtime_error("nu must be monotonically increasing.");

    // Finally, make the lookup tables;
    set_lookup_tables();
}

Dust::Dust(int _nlam, double *_nu, double *_lam, double *_kabs, 
        double *_ksca, double *_kext, double *_albedo) {
    
    nlam = _nlam;
    nu = _nu;
    lam = _lam;
    kabs = _kabs;
    ksca = _ksca;
    kext = _kext;
    albedo = _albedo;
}

Dust::~Dust() {
    delete[] temp; delete[] dkextdnu; delete[] dalbedodnu;
    delete[] planck_opacity; delete[] dplanck_opacity_dT;
    delete[] rosseland_extinction; delete[] drosseland_extinction_dT;
    delete2DArr(random_nu_CPD, ntemp, nlam);
    delete2DArr(drandom_nu_CPD_dT, ntemp-1, nlam);
    delete2DArr(random_nu_CPD_bw, ntemp, nlam);
    delete2DArr(drandom_nu_CPD_bw_dT, ntemp-1, nlam);
}

void Dust::set_lookup_tables() {
    // Create the temperature array;
    ntemp = 100;
    temp = new double[ntemp];

    for (int i = 0; i < ntemp; i++)
        temp[i] = pow(10.,-1.+i*6./(ntemp-1));

    // Calculate the derivatives of kext and albedo.
    dkextdnu = derivative(kext, nu, nlam);
    dalbedodnu = derivative(albedo, nu, nlam);

    // Calculate the Planck Mean Opacity and its derivative.
    double **tmp_planck = create2DArr(ntemp, nlam);
    for (int i = 0; i < ntemp; i++)
        for (int j = 0; j < nlam; j++)
            tmp_planck[i][j] = planck_function(nu[j], temp[i]) * kabs[j];

    planck_opacity = new double[ntemp];
    for (int i = 0; i < ntemp; i++)
        planck_opacity[i] = pi / (sigma * pow(temp[i],4)) * 
                integrate(tmp_planck[i], nu, nlam);

    dplanck_opacity_dT = derivative(planck_opacity, temp, ntemp);

    delete2DArr(tmp_planck, ntemp, nlam);

    // Calculate the Planck Mean Opacity and its derivative.
    double **tmp_ross = create2DArr(ntemp, nlam);
    for (int i = 0; i < ntemp; i++)
        for (int j = 0; j < nlam; j++)
            tmp_ross[i][j] = planck_function(nu[j], temp[i]) / kext[j];

    rosseland_extinction = new double[ntemp];
    for (int i = 0; i < ntemp; i++)
        rosseland_extinction[i] = (sigma * pow(temp[i],4) / pi) / 
                integrate(tmp_ross[i], nu, nlam);

    drosseland_extinction_dT = derivative(rosseland_extinction, temp, ntemp);

    delete2DArr(tmp_ross, ntemp, nlam);

    // Create the cumulative probability density functions for generating a
    // random nu value, for regular.
    double **tmp = create2DArr(ntemp, nlam);
    for (int i = 0; i < ntemp; i++)
        for (int j = 0; j < nlam; j++)
            tmp[i][j] = kabs[j] * planck_function(nu[j], temp[i]);

    random_nu_CPD = new double*[ntemp];
    for (int i = 0; i < ntemp; i++)
        random_nu_CPD[i] = cumulative_integrate(tmp[i], nu, nlam);

    drandom_nu_CPD_dT = derivative2D_ax0(random_nu_CPD, temp, ntemp, nlam);

    // Create the cumulative probability density functions for generating a
    // random nu value, for bw.
    for (int i = 0; i < ntemp; i++)
        for (int j = 0; j < nlam; j++)
            tmp[i][j] = kabs[j] * planck_function_derivative(nu[j], temp[i]);

    random_nu_CPD_bw = new double*[ntemp];
    for (int i = 0; i < ntemp; i++)
        random_nu_CPD_bw[i] = cumulative_integrate(tmp[i], nu, nlam);

    drandom_nu_CPD_bw_dT = derivative2D_ax0(random_nu_CPD_bw, temp, ntemp, 
            nlam);

    delete2DArr(tmp, ntemp, nlam);
}

/* Scatter a photon isotropically off of dust. */

void Dust::scatter(Photon *P) {
}

/* Absorb and then re-emit a photon from dust. */

void Dust::absorb(Photon *P, double T, bool bw) {
    double cost = -1+2*random_number();
    double sint = sqrt(1-pow(cost,2));
    double phi = 2*pi*random_number();

    P->n[0] = sint*cos(phi);
    P->n[1] = sint*sin(phi);
    P->n[2] = cost;
    P->invn[0] = 1.0/P->n[0];
    P->invn[1] = 1.0/P->n[1];
    P->invn[2] = 1.0/P->n[2];

    P->nu = random_nu(T,bw);
}

/* Calculate a random frequency for a photon. */

double Dust::random_nu(double T, bool bw) {
    double freq, CPD;

    int i = find_in_arr(T,temp,ntemp);

    double ksi = random_number();

    for (int j=1; j < nlam; j++) {
        if (bw)
            CPD = drandom_nu_CPD_bw_dT[i][j] * (T - temp[i]) + 
                random_nu_CPD_bw[i][j];
        else
            CPD = drandom_nu_CPD_dT[i][j] * (T - temp[i]) + 
                random_nu_CPD[i][j];

        if (CPD > ksi) {
            freq = random_number() * (nu[j] - nu[j-1]) + nu[j-1];
            break;
        }
    }

    return freq;
}

/* Calculate the opacity of a dust grain at a specific frequency. */

double Dust::opacity(double freq) {
    int l = find_in_arr(freq, nu, nlam);

    double opacity = dkextdnu[l]*(freq-nu[l])+kext[l];

    return opacity;
};

/* Calculate the albedo of a dust grain at a specific frequency. */

double Dust::albdo(double freq) {
    int l = find_in_arr(freq, nu, nlam);

    double albdo = dalbedodnu[l]*(freq-nu[l])+albedo[l];

    return albdo;
}

/* Calculate the Planck Mean Opacity for a dust grain at a given temperature. */

double Dust::planck_mean_opacity(double T) {
    int n = find_in_arr(T,temp,ntemp);

    double planck_mean_opacity = dplanck_opacity_dT[n]*(T-temp[n])+
        planck_opacity[n];

    return planck_mean_opacity;
}

/* Calculate the Rosseland Mean Extinction for a dust grain at a given 
 * temperature. */

double Dust::rosseland_mean_extinction(double T) {
    int n = find_in_arr(T,temp,ntemp);

    double rosseland_mean_extinction = drosseland_extinction_dT[n]*(T-temp[n])+
        rosseland_extinction[n];

    return rosseland_mean_extinction;
}
